### ISSUE TITLE
Vsync swap interval causes stutter/flicker in fastforward.

### ISSUE THREAD

**Issue by user18081972** [libretro/RetroArch#15804]
## Description

When the vsync swap interval is not 1 (unless auto=1), and fast-forwarding is used in fullscreen, the screen flickers when fastforward starts and stops.

### Expected behavior

There is no flickering when starting and stopping fastforward when vsync swap is 2 for example, and fullscreen is used.

### Actual behavior

When the vsync swap interval is not 1 (unless auto=1), and fast-forwarding is used in fullscreen, the screen flickers when fastforward starts and stops.

### Steps to reproduce the bug

1. Set Video>Synchronisation>Vsync ON + Vsync swap interval = 2
2. Video > Fullscreen mode > Start in fullscreen = on (make sure you are currently in fullscreen)
3. Fast forward in any core
4. Notice the stutter/flicker when starting and stopping fullscreen

### Bisect Results

Seems to happen in any older version of retroarch i tried.

### Version/Commit
You can find this information under Information/System Information

- RetroArch: March 15 2023 Git version b70b0de

### Environment information

- OS: Windows 11 22H2
- 120 hz monitor

**Comment by Ophidon** [comment]
What video driver are you getting this in? I'm on Windows too, and I can only replicate it with Vulkan. I've also dug at the code and have a possible fix for it in Vulkan, but I'm not yet sure the fix wouldn't cause other bugs itself.

**Comment by sonninnos** [comment]
Vulkan indeed does not enjoy vsync toggling while fullscreen. Menu forcefully enables vsync and fastforward forcefully disables vsync.

**Comment by Ophidon** [comment]
Like stated though, it doesn't occur with a swap interval of 1, just 2 or above. And 1 means vsync is on already, so that's not the culprit. 

From what I see as a possible fix: if you track the chain of events, it ends up -wanting- to change the swap interval to 0 coming from either a current value of 1 or 2, and ends up in vulkan_create_swapchain in vulkan_common.c. Which then seems to override back to 1 and decide not to recreate the swapchain after all when trying to go from the existing value of 1 to a new value of 1 via this code:

   if (     (swap_interval == 0)
         && (vk->flags & VK_DATA_FLAG_EMULATE_MAILBOX)
         && vsync)
   {
      swap_interval  =  1;
      vk->flags     |=  VK_DATA_FLAG_EMULATING_MAILBOX;
   }
   else
      vk->flags     &= ~VK_DATA_FLAG_EMULATING_MAILBOX;

   vk->flags        |= VK_DATA_FLAG_CREATED_NEW_SWAPCHAIN;

   if (       (vk->swapchain != VK_NULL_HANDLE)
         &&   (!(vk->context.flags & VK_CTX_FLAG_INVALID_SWAPCHAIN))
         &&   (vk->context.swapchain_width  == width)
         &&   (vk->context.swapchain_height == height)
         &&   ((vk->context.swap_interval    == swap_interval)))
   {
      /* Do not bother creating a swapchain redundantly. */


My quick test fix that seems to work is to change the last test there to:

   if (       (vk->swapchain != VK_NULL_HANDLE)
         && ((!(vk->context.flags & VK_CTX_FLAG_INVALID_SWAPCHAIN)) || (vk->context.swap_interval >= 1 && swap_interval >= 1))
         &&   (vk->context.swapchain_width  == width)
         &&   (vk->context.swapchain_height == height)
         &&   ((vk->context.swap_interval    == swap_interval) || (vk->context.swap_interval >= 1 && swap_interval >= 1)))
   {

Which overrides caring about swap interval changes between 1 and numbers greater than one, since swap interval > 1 doesn't actually exist in vulkan, it just gets faked by duping frames. I am far from an actual graphics programmer though to know yet if this causes other issues. Ignoring the VK_CTX_FLAG_INVALID_SWAPCHAIN that was set before it gets to this point seems a little dangerous, but that flag does still seem to get checked and cleared.

**Comment by sonninnos** [comment]
Yep, because that mailbox emulation effectively fakes and enforces vsync as swap interval 1 when it is disabled by fast-forward.
